src-self-hosted/print_info.zig

```diff
@@ -0,0 +1,184 @@
+
+    // /// Compiler version
+    // version: []const u8,
+    /// Path to lib/
+    lib_dir: []const u8,
+
+    /// Path to lib/zig/std
+    std_dir: []const u8,
+
+    /// Path to the global cache dir
+    global_cache_dir: []const u8,
+
+    const CompilerType = enum {
+        Stage1,
+        SelfHosted,
+    };
+
+    pub fn getVersionString() []const u8 {
+        // TODO: get this from build.zig somehow
+        return "0.6.0";
+    }
+
+    pub fn getCacheDir(allocator: *Allocator, compiler_type: CompilerType) ![]u8 {
+        const global_cache_dir = try getAppCacheDir(allocator, "zig");
+        defer allocator.free(global_cache_dir);
+
+        const postfix = switch (compiler_type) {
+            .SelfHosted => "self_hosted",
+            .Stage1 => "stage1",
+        };
+        return try fs.path.join(allocator, &[_][]const u8{ global_cache_dir, postfix }); // stage1 compiler uses $cache_dir/zig/stage1
+    }
+
+    // TODO: add CacheType argument here to make it return correct cache dir for stage1
+    pub fn init(allocator: *Allocator, compiler_type: CompilerType) !CompilerInfo {
+        const zig_lib_dir = try introspect.resolveZigLibDir(allocator);
+        errdefer allocator.free(zig_lib_dir);
+
+        const zig_std_dir = try fs.path.join(allocator, &[_][]const u8{ zig_lib_dir, "std" });
+        errdefer allocator.free(zig_std_dir);
+
+        const cache_dir = try CompilerInfo.getCacheDir(allocator, compiler_type);
+        errdefer allocator.free(cache_dir);
+
+        return CompilerInfo{
+            .lib_dir = zig_lib_dir,
+            .std_dir = zig_std_dir,
+            .global_cache_dir = cache_dir,
+        };
+    }
+
+    pub fn toString(self: *CompilerInfo, out_stream: var) !void {
+        inline for (@typeInfo(CompilerInfo).Struct.fields) |field| {
+            try std.fmt.format(out_stream, "{: <16}\t{: <}\n", .{ field.name, @field(self, field.name) });
+        }
+    }
+
+    pub fn toJSON(self: *CompilerInfo, out_stream: var) !void {
+        const stringifyOptions = StringifyOptions{
+            .whitespace = StringifyOptions.Whitespace{
+                // Match indentation of zig targets
+                .indent = .{ .Space = 2 },
+            },
+        };
+        try json.stringify(self, stringifyOptions, out_stream);
+    }
+
+    pub fn deinit(self: *CompilerInfo, allocator: *Allocator) void {
+        allocator.free(self.lib_dir);
+        allocator.free(self.std_dir);
+        allocator.free(self.global_cache_dir);
+    }
+};
+
+pub fn cmdInfo(allocator: *Allocator, cmd_args: []const []const u8, compiler_type: CompilerInfo.CompilerType, stdout: var) !void {
+    var info = try CompilerInfo.init(allocator, compiler_type);
+    defer info.deinit(allocator);
+
+    var bos = io.bufferedOutStream(stdout);
+    const bos_stream = bos.outStream();
+
+    var json_format = false;
+    for (cmd_args) |arg| {
+        if (mem.eql(u8, arg, "--json")) {
+            json_format = true;
```

Comment:
Maybe also have a `--text` option to set it back to non-json?

Perhaps it should be a `--format=json` argument instead?

Reply:
It's text by default, why add an additional `--text` argument?

Reply:
It's common in many command line tools so that aliases and defaults set in wrappers can be undone.

Reply:
Added a `--format` option, also changed error messages to better match other zig commands.